pkg/aws/g8scontrolplane/mutate_g8scontrolplane.go

```diff
@@ -66,89 +79,187 @@ func (m *Mutator) Mutate(request *admissionv1.AdmissionRequest) ([]mutator.Patch
+	} else if err != nil {
+		return nil, microerror.Mask(err)
+	} else {
+		// This defaulting is only done when the awscontrolplane exists
+		availabilityZones = len(awsControlPlane.Spec.AvailabilityZones)
+		patch, err = m.MutateReplicaUpdate(*g8sControlPlaneNewCR, *g8sControlPlaneOldCR, *awsControlPlane)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+		result = append(result, patch...)
 	}
 
-	var replicas int
+	patch, err = m.MutateReplicas(availabilityZones, *g8sControlPlaneNewCR, releaseVersion)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+	result = append(result, patch...)
 
-	// We only need to manipulate if replicas are not set or if its an update from single to HA master or on create
-	if g8sControlPlaneNewCR.Spec.Replicas != 0 && !isUpdateFromSingleToHA(g8sControlPlaneNewCR, g8sControlPlaneOldCR) && request.Operation != admissionv1.Create {
-		return result, nil
+	return result, nil
+}
+
+// MutateCreate is the function executed for every create webhook request.
+func (m *Mutator) MutateCreate(request *admissionv1.AdmissionRequest) ([]mutator.PatchOperation, error) {
+	var result []mutator.PatchOperation
+	var patch []mutator.PatchOperation
+	var err error
+
+	g8sControlPlaneCR := &infrastructurev1alpha2.G8sControlPlane{}
+	if _, _, err := mutator.Deserializer.Decode(request.Object.Raw, nil, g8sControlPlaneCR); err != nil {
+		return nil, microerror.Maskf(parsingFailedError, "unable to parse g8scontrol plane: %v", err)
 	}
-	infrastructureCRRef := &v1.ObjectReference{}
-	// We need to fetch the AWSControlPlane in case AZs need to be defaulted or the g8scontrolplane is just created
-	if aws.IsHAVersion(releaseVersion) || request.Operation == admissionv1.Create {
-		replicas = aws.DefaultMasterReplicas
-		update := func() error {
-			ctx := context.Background()
 
-			// We fetch the AWSControlPlane CR.
-			awsControlPlane := &infrastructurev1alpha2.AWSControlPlane{}
-			{
-				m.Log("level", "debug", "message", fmt.Sprintf("Fetching AWSControlPlane %s", g8sControlPlaneNewCR.Name))
-				err := m.k8sClient.CtrlClient().Get(ctx,
-					types.NamespacedName{Name: g8sControlPlaneNewCR.GetName(),
-						Namespace: namespace},
-					awsControlPlane)
-				if err != nil {
-					return microerror.Maskf(notFoundError, "failed to fetch AWSControlplane: %v", err)
-				}
-			}
-			// If there is an AWSControlPlane, the default replicas match the number of AZs
-			replicas = len(awsControlPlane.Spec.AvailabilityZones)
-			// If there is an AWSControlplane, we get its infrastructure reference
-			infrastructureCRRef, err = reference.GetReference(infrastructurev1alpha2scheme.Scheme, awsControlPlane)
-			if err != nil {
-				return microerror.Mask(err)
-			}
+	patch, err = m.MutateReleaseVersion(*g8sControlPlaneCR)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+	releaseVersion, err := releaseVersion(g8sControlPlaneCR, patch)
```

Comment:
We need to parse the patch from the release version label here as well, so that this doesn't fail in case it's not set on the CR yet.